documentation/design-docs/ipc-protocol.md

```diff
@@ -0,0 +1,183 @@
+    StartEventPipeTracing = 1024, // To file
+    StopEventPipeTracing,
+    CollectEventPipeTracing, // To IPC
+};
+
+struct MessageHeader
+{
+    DiagnosticMessageType RequestType;
+    uint32_t Pid;
+};
+```
+
+```
+runtime <- client : MessageHeader { CollectEventPipeTracing }
+    error? -> no response
+runtime -> client : session ID 
+runtime -> client : event stream
+
+...
+
+runtime <- client : stop command
+runtime -> client : session id and stops
+```
+
+### Proposed
+
+Connection:
+```
+runtime <- client : handshake (set length byte array that represents version)
+runtime -> client : sends handshake back as ACK
+runtime <- client : packet containing start code
+runtime -> client : ACK + potential payload
+repeat till closed
+```
+
+### Transport considerations
+
+* All data will be sent in the endianess of the machine
+* All data will be sent in a packet containing a header of information concerning the contents
+
+### Handshake
+
+#### Requirements:
+* simple - easy to code
+* fast - no complex structures/serialization
+* extensible - leaves room to change resulting behavior
+
+The client will initiate the connection by sending a byte array containing the ASCII codes for the version string of the protocol it wishes to communicate with.
+
+> There will only be one for now, but it will give us the ability to extend this and potentially have _other_ protocols we can speak if we'd like.
+
+```c
+// "EVENTPIPE_V1"
+char[15] handshake = { 0x00, 0x00, 0x00, 0x45, 0x56, 0x45, 0x4e, 0x54, 0x50, 0x49, 0x50, 0x45, 0x5f, 0x56, 0x31 };
+```
+
+The `Accept` function in the runtime will immediately try to read 15 bytes of data and then use that determine what protocol we are speaking.  Once it has been determined, the runtime will respond with that handshake.
+
+#### Errors
+* runtime: 
+  * read failure -> end connection; no response
+  * unknown protocol -> ERROR response // TODO: flesh out errors
+* Client:
+  * no response -> simply have a timeout
+  * unkown protocol error -> fix your client
+
+### Headers
+
+#### Requirements:
+* small - shouldn't affect write times or read times
+* simple - flat and blittable
+* descriptive - self describing
+* extensible - ability for us to add funcitonality without necessarily incrementing protocol version
+
+Every packet will start with a header and every header will start with a size.  This makes it easy to read the data directly out of the stream no matter the size or contents of the packet.
+
+> Taking inspiration from Mono's Soft-Debugger protocol, I am splitting command codes into command sets and commands, allowing command codes to be reused between command sets, e.g., 0x01,0x01 and 0x02,0x01 would be different commands despite sharing the same explicit value for the command code.
+
+```c
+// size = 4 + 1 + 1 + 1 = 7 bytes
+struct header
+{
+    uint16 size; // size of packet = size of header + payload
+    char   flag; // command or response
+    char   command_set;
+    char   command;
```

Comment:
By splitting this into two separate bytes, it means the main diagnostics server only needs to concern itself with the `command_set` and then the various controllers, e.g., EventPipe controller can handle the individual commands.  Granted, this only ever affects the first communication since subsequent interactions will be handled by the controller.

for example:
```c
// in diagnostics server
enum command_sets : char
{
    EventPipe = 0x01,
    // hypotheticals
    Debugger = 0x02,
    Other = 0x03,
    ...
}

diag server receives connection
diag server checks command code (in this example it is 0x01)
forwards connection to EventPipeController

// in EventPipeController
enum commands : char
{
    start = 0x01,
    stop = 0x02,
    ...
}
```

This structure allows to keep information separated and loosely coupled.  The parent level enum for command sets now only contains information about possible controllers and doesn't need to know what they are capable of doing.

Reply:
I'm not pushing back against having additional semantic structure inside the id, I was only suggesting that it doesn't need to be codified as part of the protocol. For example if the protocol defines a single 16 bit command, it would be perfectly legal implementation for us to assign the ids:

    0x0001 = EventPipeCommand1
    0x0002 = EventPipeCommand2
    0x0101 = DumpCommand1
    0x0201 = ProfilerCommand1
    …

I think of it the same way that IP protocols declare addresses as 4 byte integers. There is tons of structure inside the IP address that routers make use of, but that doesn't prevent it from being abstracted on the wire.

Reply:
Ah, that makes sense.  We could just have it be part of the semantics of use that the top two nibbles are the command set and the bottom two are the command.  Getting the information out is as simple as `command & 0x1100` and `command & 0x0011`.  I think the server code should maybe separate them so that the commands available aren't as tightly coupled with the diagnostics server.  I think it would be cleaner to have all the commands a controller can receive defined in the controller rather than in the parent server.